tests/src/test/scala/akka/kafka/scaladsl/RebalanceExtSpec.scala

```diff
@@ -0,0 +1,427 @@
+    Consumer
+      .committablePartitionedSource(
+        consumerSettings.withClientId(clientId),
+        subscription
+          .withPartitionAssignmentHandler(assignmentHandler(clientId))
+          .withRebalanceListener(rebalanceListener)
+      )
+      .map {
+        case (topicPartition, topicPartitionStream) =>
+          log.debug(s"$logPrefix::Consuming partitioned source clientId: $clientId, for tp: $topicPartition")
+          val innerStream: Source[ConsumerMessage.CommittableOffsetBatch, NotUsed] = topicPartitionStream
+            .via(sharedKillSwitch.flow)
+            .via(
+              businessFlow(
+                clientId,
+                perPartitionMessageCount,
+                topicPartitionFutureMap,
+                messageStoreAndAck,
+                sharedKillSwitch,
+                logPrefix
+              )
+            )
+            .via(Committer.batchFlow(committerDefaults.withMaxBatch(1)))
+          innerStream.runWith(Sink.ignore)
+      }
+      .via(sharedKillSwitch.flow)
+      .toMat(Sink.seq)(Keep.both)
+      .run()
+  }
+
+  // create topic-partition map
+  def createTopicMapsAndPublishMessages(
+      topicCount: Int,
+      partitionCount: Int,
+      perPartitionMessageCount: Int
+  ): (Map[String, MSet[TopicPartition]],
+      Map[String, Promise[Done]],
+      Map[Int, String],
+      Set[String],
+      Map[Int, String],
+      Seq[Future[Done]],
+      Map[Int, (AtomicInteger, Promise[Done], Promise[Done])]) = {
+    var topicMap = Map[String, MSet[TopicPartition]]()
+    var topicPartitionFutureMap = Map[String, Promise[Done]]()
+    var topicIdxMap = Map[Int, String]()
+    (1 to topicCount).foreach(topicIdx => {
+      val topic1 = createTopic(topicIdx, partitions = partitionCount)
+      log.debug(s"created topic topic1=$topic1")
+      topicIdxMap = topicIdxMap.updated(topicIdx, topic1)
+      topicMap = topicMap.updated(topic1, MSet[TopicPartition]())
+      (1 to partitionCount).foreach(partitionIdx => {
+        val tp1 = new TopicPartition(topic1, partitionIdx - 1)
+        topicMap(topic1).add(tp1)
+        topicPartitionFutureMap = topicPartitionFutureMap.updated(s"$topic1-${partitionIdx - 1}", Promise[Done])
+      })
+    })
+    val (msgTpMap, producers, messageStoreAndAck) =
+      publishMessages(topicCount, partitionCount, perPartitionMessageCount, topicIdxMap)
+    (topicMap, topicPartitionFutureMap, topicIdxMap, topicMap.keySet, msgTpMap, producers, messageStoreAndAck)
+  }
+
+  // publish a numeric series tagged messages across all topic-partitions
+  // e.g., for 18 messages:  messageId(1 to 9) => topic-1-1-0; messageId(10 to 18) => topic-1-1-1
+  def publishMessages(
+      topicCount: Int,
+      partitionCount: Int,
+      perPartitionMessageCount: Int,
+      topicIdxMap: Map[Int, String]
+  ): (Map[Int, String], Seq[Future[Done]], Map[Int, (AtomicInteger, Promise[Done], Promise[Done])]) = {
```

Comment:
Introduce a type for this return signature.

Reply:
done